File: everything_parsing/19-recursive-descent/js/derivation.mjs
// Replays a production list as a LEFTMOST derivation (mirror of java/Derivation.java).
// Start at [E]; for each production, replace the leftmost nonterminal with its
// RHS. If the list is a recursive-descent firing order, every LHS matches the
// leftmost nonterminal and the final form is the input's token kinds.

const NONTERMINALS = new Set(["E", "E'", "T", "T'", "F"]);

export function replay(rules) {
  let form = ["E"];
  const steps = [{ form: join(form), rule: null }];

  for (const rule of rules) {
    const i = leftmostNonterminal(form);
    if (i < 0) throw new Error(`form has no nonterminal but rule remains: ${rule}`);

    const [lhs, rhsText] = rule.split(" -> ");
    if (form[i] !== lhs.trim())
      throw new Error(
        `rule '${rule}' expands ${lhs.trim()} but the leftmost nonterminal is ${form[i]} ` +
        `— the derivation is not leftmost`);

    const rhs = rhsText.trim().split(/\s+/).filter(s => s !== "epsilon");
    form = [...form.slice(0, i), ...rhs, ...form.slice(i + 1)];
    steps.push({ form: join(form), rule });
  }
  return steps;
}

export function sentence(rules) {
  const s = replay(rules);
  return s[s.length - 1].form;
}

function leftmostNonterminal(form) {
  for (let i = 0; i < form.length; i++) if (NONTERMINALS.has(form[i])) return i;
  return -1;
}

function join(form) {
  return form.length === 0 ? "epsilon" : form.join(" ");
}

File: everything_parsing/19-recursive-descent/java/Derivation.java
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Replays a list of productions as a LEFTMOST derivation.
 *
 * Start at the sentential form [E]. For each production in the list, find the
 * leftmost nonterminal in the current form and replace it with the production's
 * right-hand side. If the list really is the firing order of a recursive-descent
 * parse, then (a) every production's left-hand side matches the leftmost
 * nonterminal at that step — never anything else — and (b) the final form is
 * exactly the input's token kinds. That is the proof that "the call stack is the
 * leftmost derivation".
 */
public final class Derivation {

    private static final Set<String> NONTERMINALS = Set.of("E", "E'", "T", "T'", "F");

    public record Step(String form, String rule) {}

    public static List<Step> replay(List<String> rules) {
        List<String> form = new ArrayList<>(List.of("E"));
        List<Step> steps = new ArrayList<>();
        steps.add(new Step(join(form), null));

        for (String rule : rules) {
            int i = leftmostNonterminal(form);
            if (i < 0)
                throw new IllegalStateException("form has no nonterminal but rule remains: " + rule);

            String[] halves = rule.split(" -> ", 2);
            String lhs = halves[0].trim();
            if (!form.get(i).equals(lhs))
                throw new IllegalStateException(
                    "rule '" + rule + "' expands " + lhs + " but the leftmost nonterminal is " + form.get(i)
                    + " — the derivation is not leftmost");

            List<String> rhs = new ArrayList<>();
            for (String sym : halves[1].trim().split("\\s+"))
                if (!sym.equals("epsilon")) rhs.add(sym);

            form.remove(i);
            form.addAll(i, rhs);
            steps.add(new Step(join(form), rule));
        }
        return steps;
    }

    /** Just the last sentential form — the sentence the derivation produces. */
    public static String sentence(List<String> rules) {
        List<Step> s = replay(rules);
        return s.get(s.size() - 1).form();
    }

    private static int leftmostNonterminal(List<String> form) {
        for (int i = 0; i < form.size(); i++)
            if (NONTERMINALS.contains(form.get(i))) return i;
        return -1;
    }

    private static String join(List<String> form) {
        return form.isEmpty() ? "epsilon" : String.join(" ", form);
    }
}
